Range repair tasks

Deexie opened the pull request:
Repair tasks spanning over a range are introduced.

The progress of repair tasks is counted basing on the number
of ranges that were processed. Completed ranges in shard 
and top level tasks are gathered recursively from their children.

Deexie commented:
@bhalevy review please

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3528/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3527/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3527/

failures in task manager test

bhalevy commented on repair/repair.cc:
```diff
@@ -938,6 +938,11 @@ static future<tasks::task_manager::task_ptr> start_repair_task(tasks::task_manag
     co_return task;
 }
 
+future<> range_repair_task_impl::run() {
```
Please provide a more verbose git commit message to explain the motivation behind it.

bhalevy commented on repair/repair_task.hh:
```diff
@@ -148,6 +148,46 @@ protected:
+            std::string table,
+            std::string type,
+            repair_uniq_id parent_id,
+            shard_repair_task_impl& shard_task,
```
Are we sure it won't move?

bhalevy commented on repair/repair.cc:
```diff
@@ -640,53 +640,61 @@ future<> shard_repair_task_impl::repair_range(const dht::token_range& range, ::t
             rlogger.error("repair[{}]: Repair {} out of {} ranges, shard={}, keyspace={}, table={}, range={}, peers={}, live_peers={}, status={}",
-                    id.uuid(), ranges_index, ranges_size(), id.shard(), _status.keyspace, table_names(), range, neighbors, live_neighbors, status);
-            abort_repair_info();
+                    _shard_task.id.uuid(), _ranges_index, _shard_task.ranges_size(), _shard_task.id.shard(), _status.keyspace, _shard_task.table_names(), _range, _neighbors, _live_neighbors, status);
```
it seems like a preexisting bug that we print `table_names()` rather than the particular table the task is working on, no?

bhalevy replied:
cc @asias

asias replied:
Yes, after 3dc9a81d02a26d021ad1954d57b574ad4899561d, we now only repair one table per range instead of all tables per range.

asias replied:
Yes, since 3dc9a81d02a26d021ad1954d57b574ad4899561d, it repairs one table per range instead of all tables per range.

bhalevy commented on repair/repair.cc:
```diff
@@ -640,53 +640,61 @@ future<> shard_repair_task_impl::repair_range(const dht::token_range& range, ::t
-        if (it == live_neighbors.end()) {
-            nr_failed_ranges++;
+
+    auto task_impl_ptr = std::make_unique<range_repair_task_impl>(rs.get_repair_module().shared_from_this(), tasks::task_id::create_random_id(), _status.keyspace, table_name, format("{}", reason), get_repair_uniq_id(), *this, range, table_id, std::move(neighbors), std::move(mandatory_neighbors), std::move(live_neighbors), ranges_index);
```
nit: std::move(table_name)

bhalevy commented on repair/repair_task.hh:
```diff
@@ -17,6 +17,17 @@ public:
+        while (i < _children.size()) {
+            progress.completed += co_await gather_child_completed(i);
+            ++i;
+        }
```
this looks like a for loop to me.
How about:
```c__
    tasks::task_manager::task::progress progress{};
    for (unsigned i = 0; i < _children.size(); ++i) {
        progress.completed += co_await gather_child_completed(i);
    }
```

or even better, have `gather_child_completed` accept a `const foreign_task_ptr&`, and then the loop will be even simpler:
```c++
    for (const auto& child : _children) {
        progress.completed += co_await gather_child_completed(child);
    }
```

Deexie replied:
Child tasks may still be pushed to `_children` in the meantime. Thus, foreach shouldn't be used. 

I will change the while loop to for and leave a comment about the above.

bhalevy replied:
still, sorry, but I don't like gather_child_completed given an index rather than a ref to child.
That said, if it has no other use, let's just open code it here.  I don't think it warrants a method (unless again if it would be reused elsewhere)

bhalevy commented on test/rest_api/rest_util.py:
```diff
@@ -3,6 +3,8 @@
 from contextlib import contextmanager
 
+long_time = 1000000000
+
```
I don't think this belongs to rest_util.
Just define it locally in the function that uses it.
There's nothing special in this constant.

Deexie commented:
Sorry, fixing

Deexie commented:
- skip repair task in release mode
- edited commit message of the `repair: add range_repair_task_impl`
- used std::move on table_name
- used a name of a particular table instead of all the tables
- replaced while with for loop in repair_task_impl::get_progress()
- deleted `long_time` from `rest_util.py`
- deleted move and copy ctors from task::impl
- rebased on master

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3556/

bhalevy commented on tasks/task_manager.hh:
```diff
@@ -100,6 +100,8 @@ public:
         public:
             impl(module_ptr module, task_id id, uint64_t sequence_number, std::string keyspace, std::string table, std::string entity, task_id parent_id) noexcept;
+            impl(const impl&) = delete;
+            impl(impl&&) = delete;
```
Please add a comment explaining why, documenting that task_manager::task::impl is always created as a unique_ptr hence it need not move or be copied.

bhalevy commented on repair/repair.cc:
```diff
@@ -950,9 +953,9 @@ future<> shard_repair_task_impl::do_repair_ranges() {
-        co_await coroutine::parallel_for_each(ranges, [this, table_id] (auto&& range) {
-            return with_semaphore(rs.get_repair_module().range_parallelism_semaphore(), 1, [this, &range, table_id] {
-                return repair_range(range, table_id).then([this] {
+        co_await coroutine::parallel_for_each(ranges, [this, table_id, table_name] (auto&& range) {
```
since this runs in a coroutine, we can capture `&table_name` by reference and copy it only when calling `repair_range`.

bhalevy commented on repair/repair.cc:
```diff
@@ -652,7 +652,7 @@ future<> range_repair_task_impl::run() {
             auto status = format("failed: mandatory neighbor={} is not alive", node);
             rlogger.error("repair[{}]: Repair {} out of {} ranges, shard={}, keyspace={}, table={}, range={}, peers={}, live_peers={}, status={}",
-                    _shard_task.id.uuid(), _ranges_index, _shard_task.ranges_size(), _shard_task.id.shard(), _status.keyspace, _shard_task.table_names(), _range, _neighbors, _live_neighbors, status);
+                    _shard_task.id.uuid(), _ranges_index, _shard_task.ranges_size(), _shard_task.id.shard(), _status.keyspace, _status.table, _range, _neighbors, _live_neighbors, status);
```
great, thanks for cleaning this up.

bhalevy replied:
@asias please review too since this pre-existed before this PR.

bhalevy commented on repair/repair_task.hh:
```diff
@@ -193,6 +193,14 @@ public:
+    virtual future<tasks::task_manager::task::progress> get_progress() const override {
+        auto state = _status.state;
+        co_return tasks::task_manager::task::progress{
+            .completed = state == tasks::task_manager::task_state::done || state == tasks::task_manager::task_state::failed,
```
don't we have a method to check the task_state for completion?
If not, let's add one since the logic is generic and can very well be used by other task types.

bhalevy replied:
Maybe this could even be the generic implementation of this virtual function in the base class.

bhalevy reviewed (changes requested):
I had some minor comments, but we're getting there :)

asias commented on repair/repair.cc:
```diff
@@ -638,53 +638,61 @@ future<> shard_repair_task_impl::repair_range(const dht::token_range& range, ::t
-        if (it == live_neighbors.end()) {
-            nr_failed_ranges++;
+
+    auto task_impl_ptr = std::make_unique<range_repair_task_impl>(rs.get_repair_module().shared_from_this(), tasks::task_id::create_random_id(), _status.keyspace, std::move(table_name), get_repair_uniq_id(), *this, range, table_id, std::move(neighbors), std::move(mandatory_neighbors), std::move(live_neighbors), ranges_index, _reason);
```
Too many paramters. As a follow up, we can define a class for those parameters and pass a single object to range_repair_task_impl.

asias commented on repair/repair.cc:
```diff
@@ -1311,6 +1319,13 @@ future<> repair_service::sync_data_using_repair(
+    auto module = dynamic_pointer_cast<repair_module>(_module);
+    auto& db = module->get_repair_service().get_db().local();
+    auto cfs = list_column_families(db, _status.keyspace);
+    return _ranges.size() * cfs.size() * smp::count;
```
In data_sync_repair_task_impl::run, we have the logic to list the tables we need to repair for the keyspace. Can we remove the duplication in data_sync_repair_task_impl::get_total() ?

Deexie replied:
The main reason why I did it like this is that before this line is reached:
https://github.com/scylladb/scylladb/blob/d7f56d736764a37c36d37c2b882aef730f014d59/repair/repair.cc#L1340
user can already request task status and `cfs.size()` is needed for calculating total progress. 

If the main problem here is time consumed by `list_column_families()` then it can be done before the task is created and passed to `data_sync_repair_task_impl` constructor. I am not sure if this is ok, since list of cf sounds like something pretty dynamic, is it? Oh, if it is, then current get_total would return different result each time its called...

If it's about the whole:
```
auto module = dynamic_pointer_cast<repair_module>(_module);
auto& db = module->get_repair_service().get_db().local();
```
Then we could set the `cfs.size()` with setter and return 0.0/1.0 progress if it hasn't been set yet. This would slightly miss the "progress_units="ranges" but I think it's not that bad. Would that be ok?

asias replied:
Yes, I know that the number of tables to be repaired is only called inside data_sync_repair_task_impl::run. It is important to get a consistent number of tables in get_total() as the task will will repairs in data_sync_repair_task_impl::run.

We can return 0% if  we haven't reached the run() where the correct get_total is calculated. It is ok to "miss" some progress since user won't be able to query all the progresses anyway. 

We can set the total size in run(), so get_total() will be faster too.

Deexie commented:
Turned to draft because of std::bad_array_new_length being thrown after changes. Investigating.

Deexie commented:
- rebased on master
- added a comment to impl about deleting copy and move constructor
- capture table_name by reference
- moved range_repair_task_impl::get_progress() implementation to impl::get_progress()
- modified `data_sync_repair_task_impl::get_total()`: currently we set `_cfs_count` in `run()` and return progress depending on whether `_cfs_count` is set
- opencoded gather_child_completed
The std::bad_array_new_length still exists. I just wanted to push the version with resolved comments.

Deexie commented:
- rebased on master
- used module::make_and_start_task()

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4319/

bhalevy commented on repair/repair.cc:
```diff
@@ -584,68 +584,76 @@ repair_neighbors shard_repair_task_impl::get_repair_neighbors(const dht::token_r
 // Repair a single local range, multiple column families.
 // Comparable to RepairSession in Origin
-future<> shard_repair_task_impl::repair_range(const dht::token_range& range, ::table_id table_id) {
+future<> shard_repair_task_impl::repair_range(const dht::token_range& range, ::table_id table_id, sstring table_name) {
```
Why do we need to pass `table_name`?
`shard_repair_task_impl` keeps reference to `db` and already does:
https://github.com/scylladb/scylladb/blob/941407b905a60923e0b0ac1e6b0399cac2d67b79/repair/repair.cc#L635

bhalevy replied:
@Deexie, `table_name` is still passed here, but it is not used.

bhalevy reviewed (commented):
Other than the extraneous `table_name` parameter looks good to me.

Deexie commented:
- delete table name from range task constructor
- replace `cf` with `_status.table` in range_repair_task_impl::run()
- rebase on master
	- use task_manager_utils.py in test
    - allow test_repair_task.py in release mode

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4517/

Deexie commented:
- delete `table_name` argument from `shard_repair_task_impl::repair_range`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4552/

Deexie commented:
- added table_name deleted accidentally from log message

bhalevy commented:
> * added table_name deleted accidentally from log message

Thanks.  Please make a habit of running ./test.py --mode=dev locally before updating the PR branch

Deexie commented:
> Thanks. Please make a habit of running ./test.py --mode=dev locally before updating the PR branch

I do. Failed tests were from dtest, which I am not sure if should be run locally after each change

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4560/

Deexie commented:
- rebase

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/86/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/115/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2441/

Deexie commented:
- rewrite & rebase

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2447/

Deexie commented:
Closing as it's abandoned and replaced
